CRM/Mailing/Event/BAO/Unsubscribe.php

```diff
@@ -140,75 +140,60 @@ public static function &unsub_from_mailing($job_id, $queue_id, $hash, $return =
+    // tests, it's the variant 'A' one.
+    $relevant_mailing_id = $mailing_id;
+
+    // Special case for AB Tests:
+    if (in_array($mailing_type, ['experiement', 'winner'])) {
+      // The mailing belongs to an AB test.
+      // See if we can find an AB test where this is variant B.
+      $mailing_id_a = CRM_Core_DAO::getFieldValue('CRM_Mailing_DAO_MailingAB', mailing_id, 'mailing_id_a', 'mailing_id_b');
+      if (!empty($mailing_id_a)) {
+        // OK, we were given mailing B and we looked up variant A which is the relevant one.
+        $relevant_mailing_id = $mailing_id_a;
+      }
+      else {
+        // No, it wasn't variant B, let's see if we can find an AB test where
+        // the given mailing was the winner (C).
         $mailing_id_a = CRM_Core_DAO::getFieldValue('CRM_Mailing_DAO_MailingAB', $mailing_id, 'mailing_id_a', 'mailing_id_c');
-        $field = 'mailing_id_c';
+        if (!empty($mailing_id_a)) {
+          // OK, this was the winner and we looked up variant A which is the relevant one.
+          $relevant_mailing_id = $mailing_id_a;
+        }
+        // (otherwise we were passed in variant A so we already have the relevant_mailing_id correct already.)
       }
-      $jobJoin = "INNER JOIN $ab ON $ab.mailing_id_a = $mg.mailing_id
-        INNER JOIN $job ON $job.mailing_id = $ab.$field";
-      $entity = CRM_Core_DAO::getFieldValue('CRM_Mailing_DAO_MailingGroup', $mailing_id_a, 'entity_table', 'mailing_id');
-    }
-    else {
-      $jobJoin = "INNER JOIN  $job ON      $job.mailing_id = $mg.mailing_id";
     }
 
-    $groupClause = '';
-    if ($entity == $group) {
-      $groupClause = "AND $group.is_hidden = 0";
-    }
-
-    $do->query("
-            SELECT      $mg.entity_table as entity_table,
-                        $mg.entity_id as entity_id,
-                        $mg.group_type as group_type
-            FROM        $mg
-            $jobJoin
-            INNER JOIN  $entity
-                ON      $mg.entity_id = $entity.id
-            WHERE       $job.id = " . CRM_Utils_Type::escape($job_id, 'Integer') . "
-                AND     $mg.group_type IN ('Include', 'Base') $groupClause"
-    );
-
-    // Make a list of groups and a list of prior mailings that received
-    // this mailing.
+    // Make a list of groups and a list of prior mailings that received this
+    // mailing.  Nb. the 'Base' group is called the 'Unsubscribe group' in the
+    // UI.
+    // Just to definitely make it SQL safe.
+    $relevant_mailing_id = (int) $relevant_mailing_id;
+    $do = CRM_Core_DAO::executeQuery(
+      "SELECT entity_table, entity_id, group_type
+        FROM civicrm_mailing_group
+       WHERE mailing_id = $relevant_mailing_id
+         AND group_type IN ('Include', 'Base')");
 
     $groups = [];
     $base_groups = [];
     $mailings = [];
 
     while ($do->fetch()) {
-      if ($do->entity_table == $group) {
+      if ($do->entity_table == 'civicrm_group') {
```

Comment:
this strikes me a potentially problematic

Reply:
@seamuslee001 why - we save the table as entity_table in that field not the internationalised version don't we?

Reply:
@eileenmcnaughton nope its the internationalised version posting comment shortly

Reply:
@seamuslee001 ug - we shouldn't - that isn't a pattern anywhere else. We should fix on upgrade

Reply:
its been like that for ages, hence why we have put in mulitlingual testing into api/v3/mailing etc. it would be better to sort it long term but that will take some hand holding as there are a number of places in the mailing code where it assumes the internationalised table name is stored

Reply:
Wow - it's an ongoing problem isn't it :-(

Reply:
Still for this PR we can make all the other changes except for that & it will still clean it up a lot

Reply:
Yep agreed and if we make the test changes as per my comment below then we will have test coverage in single and mulitlingual mode for whenever we get around to doing the cleanup work

Reply:
Does it make sense for you to do a PR against this branch?

Reply:
done here @eileenmcnaughton https://github.com/artfulrobot/civicrm-core/pull/1